### PR TITLE
Allow decomposing composite gates by name

### LINKED ISSUE #9136: Circuit.decompose doesn't decompose XXPlusYYGate
### Environment

- **Qiskit Terra version**: dfbc738be57619b503b931a2d0741433a1bbe21b
- **Python version**: 3.10.8
- **Operating system**: Arch Linux

### What is happening?

See title.

### How can we reproduce the issue?

```python
from qiskit.circuit import QuantumCircuit
from qiskit.circuit.library import XXPlusYYGate

circuit = QuantumCircuit(2)
circuit.append(XXPlusYYGate(0.1, 0.1), [0, 1])
print(circuit.decompose(["xx_plus_yy"]))
```
```
     ┌───────────────────┐
q_0: ┤0                  ├
     │  {XX+YY}(0.1,0.1) │
q_1: ┤1                  ├
     └───────────────────┘
```

### What should happen?

It should decompose.

### Any suggestions?

_No response_

### PR DESCRIPTION
**Pull request by enavarro51**
### Summary

Fixes #9136

### Details and comments

This fixes a situation where passing the name of a composite gate to `QuantumCircuit.decompose` would result in the gate not decomposing if there was a label on the gate.

### CODE REVIEW

**Review by Cryoris — COMMENTED** [review]
I left some preliminary comments, but I'm not sure this behavior is what we want. The idea of decomposing by name was to decompose according to what was drawn in the circuit drawer, but now that's not true anymore. For example, with the circuit of the test:
```
      ┌────────┐               ┌───┐┌─────────────┐
q0_0: ┤0       ├────────────■──┤ H ├┤0            ├
      │        │            │  └───┘│  circuit-93 │
q0_1: ┤1 gate1 ├────────────■───────┤1            ├
      │        │┌────────┐  │       └─────────────┘
q0_2: ┤2       ├┤0       ├──■──────────────────────
      └────────┘│        │  │
q0_3: ──────────┤1 gate2 ├──■──────────────────────
                │        │┌─┴─┐
q0_4: ──────────┤2       ├┤ X ├────────────────────
                └────────┘└───┘
```
specifying `"circuit-*"` should only decompose the top-right box, but now it also decomposes `gate1` and `gate2`, to
```
      ┌───┐               ┌───┐┌───┐
q0_0: ┤ H ├────────────■──┤ H ├┤ X ├
      ├───┤            │  └───┘└───┘
q0_1: ┤ T ├────────────■────────────
      ├───┤┌───┐       │
q0_2: ┤ X ├┤ H ├──■────■────────────
      └───┘└───┘┌─┴─┐  │
q0_3: ──────────┤ X ├──■────────────
      ┌───┐     └───┘┌─┴─┐
q0_4: ┤ X ├──────────┤ X ├──────────
      └───┘          └───┘
```
but this is not how it was originally designed. If we _do_ want that we can change it, but I just want to make sure that this right now is a breaking change.

In the original design it was intended that `XXPlusYYGate` was indeed only decomposed by 
```
circuit.decompose("{XX+YY}")
```
and not by tagging `"xx_plus_yy"`.

**Comment by jakelishman** [comment]
Julien: I wasn't aware that was the intention from the PR that added it, but even learning it now, it feels a bit odd to me.  There's no inherent link between `decompose` and the drawers, and no link stated in the documentation, or anything like that.  `decompose` is also used by people outside the context of the drawers (such as a quick way to open up some gates into more fundamental definitions), so it doesn't seem like a great idea to tie them too tightly.

The only documentation that exists on using strings as the decomposition is in the `Decomposer` class (which no-one will see), which says "`gates_to_decompose`: ... identified by gate label, name or type."  To me, that suggests both label and name should work.  At the very least, I think we'll need to update the documentation of `Decomposer` (and `QuantumCircuit.decompose`) if it's meant to be another way, but I would argue in favour of what Edwin's done here - you can see from the linked issue the current way is already causing confusion for gates that have to define a `label` for other reasons relating to the drawer.

File: qiskit/transpiler/passes/basis/decompose.py [hunk 1]
```diff
@@ -121,15 +119,12 @@ def _should_decompose(self, node) -> bool:
         strings_list = [s for s in gates if isinstance(s, str)]
         gate_type_list = [g for g in gates if isinstance(g, type)]
 
-        if hasattr(node.op, "label") and node.op.label is not None:
-            has_label = True
-
-        if has_label and (  # check if label or label wildcard is given
+        if node.op.label and (  # check if label or label wildcard is given
```
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/transpiler/passes/basis/decompose.py:122]: It was actually intentional as I was trying to pick up `!= ""` as well, but I can put that in explicitly.

File: qiskit/circuit/library/standard_gates/xx_minus_yy.py [hunk 2]
```diff
@@ -95,7 +95,7 @@ def __init__(
         self,
         theta: ParameterValueType,
         beta: ParameterValueType = 0,
-        label: Optional[str] = "{XX-YY}",
+        label: Optional[str] = "(XX-YY)",
```
[the remark this answers is not in the document]
Comment (enavarro51) [reply qiskit/circuit/library/standard_gates/xx_minus_yy.py:98]: Will do.

**Comment by enavarro51** [comment]
@Cryoris, @jakelishman. I see both sides of this, so just let me know which way to go. I also noticed that `gate` has been long deprecated in that method, so I'll remove it. I can add some docs to `QuantumCircuit.decompose` to better explain the behavior you decide on.

**Comment by Cryoris** [comment]
Yeah I agree that the mix right now is confusing, it seems we were trying to solve a different problem before. But just decomposing if either of the label or name fits still seems imprecise.  Would it be a cleaner solution to separately have either (1) gate type and name (both sort of are the type) and (2) labels? That could be done by eg adding "labels_to_decompose" or something similar. 

We could also come up with a different solution for checking what gates are made up of in the drawer, but that would probably be very helpful to have (which is why we made decompose do it before 🙂)

**Comment by enavarro51** [comment]
@Cryoris Adding the 'labels_to_decompose' could be a good solution for the current situation. It eliminates having to decide if 'name' or 'label' has some kind of priority. Only issue is in the short-term if someone is currently using labels in 'gates_to_decompose', how do we handle that. Do a deprecation for it? @jakelishman Let me know if this seems reasonable to you and I can add it in.

In regard to showing gates composed of circuits in the drawers, I'm currently working on showing the control flow circuit clauses in the 'mpl' drawer. It's not a simple problem, but if I get that working, the same structure could be used to show any composite gate as a circuit, perhaps with an option for the user.

**Comment by jakelishman** [comment]
_Technically_ we should have a deprecation, but this decomposition by string is sufficiently new (and the behaviour sufficiently unexpected, as evidenced by the ticket) that we can say the previous behaviour of `gates_to_decompose` decomposing labels was a "bug", and that this `labels_to_decompose` is a new "feature".  It's playing a _little_ fast and loose with the deprecation policy, but I think the impact on users is likely to be very minimal, tending towards positive.  I'm happy to move ahead with the two separate arguments.

Those changes to the drawer sound great, Edwin, thanks!

**Comment by enavarro51** [comment]
In thinking (overthinking) this further, I don't believe the `labels_to_decompose` is the answer. If we look at the original problem, the user was trying to use `circuit-*` to decompose all the composite gates and only the composite gates. Because composite gates generate a name with a number that can change from one run to the next, there is no way to specify a single composite gate using the name. This is what I think should happen for these entries in `gates_to_decompose`
- "circuit-*" - decompose all composite gates
- "my_gate_label*" - decompose all gates that match
- "my_gate_label3" - decompose only that gate

Bottom line is the user can only specify a single or subset of all composite gates by using labels. This can be explained in detail in the docs. So leave the behavior as it is in this PR, with some update to the docs.

**Comment by jakelishman** [comment]
Edwin: this seems to miss the issue that actually caused this PR to be opened, which was that we were surprised that you can't use the `gate_name` for `XXPlusYYGate` (`xx_plus_yy`) to decompose those gates, because for other unrelated drawing reasons, that gate class has to define different label.  That's surprising behaviour, and the goal of this is to ensure that gate name strings (which are a first-class citizen of the Qiskit data model) can be used to select gates, and `label` (which is largely a drawer-only feature, and one we're not even convinced will stay) doesn't prevent that.

`Decomposer` and `decompose` aren't drawer-specific, and never were (for example, see [the PR that introduced `Decomposer`](https://github.com/Qiskit/qiskit-terra/pull/1487)).

**Review by jakelishman — APPROVED** [review]
Thanks Edwin - I think this now agrees with what we'd decided.  I'll merge this now.
